### DISCUSSION TITLE
Excessive RAM consumption

### DISCUSSION THREAD

**Discussion by bernejorge** [FlowiseAI/Flowise#3733] [Q&A] [answered]
I have a problem with excessive RAM consumption?
I'm using a sequential agent.
Any ideas? Has this happened to anyone?

![image](https://github.com/user-attachments/assets/5f8914d3-970e-4720-a0e2-66035db895e2)

**Comment by severfire** [comment]
does it grow with every api call?

**Reply by bernejorge** [reply]
yes

**Comment by bitsnaps** [comment]
I've noticed the same thing when uploading txt/csv file, it's relatively small compared to when using custom python with chromaDB code, also knowing that everything is running on huggingface with 16Gb of vRAM and relaying on Groq API not a local model!
Suggested solution: provide a way to consult the log, I used LangSmith for logging but couldn't spot the issue!

**Comment by bernejorge** [comment, marked as answer]
I detected that graphs are added to a pool, but never removed, except when the prediction is aborted. The pool accumulates the chatflow by session id. Furthermore, this pool is not used as a cache since the nodes need to be reprocessed for each request to resolve the interpolated variables with double braces{{ }}.
Also, rebuilding the graph for each request takes a lot of memory and CPU resources

![image](https://github.com/user-attachments/assets/a6c48fe0-f618-494a-8101-9f6460a893f2)
![image](https://github.com/user-attachments/assets/24d95ae4-5a9e-49f8-a5ee-d6481984595a)

I solved the problem by creating a cache and forgoing using variable interpolation with double braces that must be resolved before creating the langgraph node and using state variables instead.
I did stress testing and it is now working in production.
So far so good.

![image](https://github.com/user-attachments/assets/d08b09ac-9253-4418-8751-9b45473b0ba0)
![image](https://github.com/user-attachments/assets/6f09c816-b9a2-4935-a55d-53054dabdd27)

**Reply by vsalbuq** [reply]
It would be nice to have a PR. 

I'm experiencing the same problem here.

**Reply by bernejorge** [reply]
Try this docker image:
docker pull bernejorge/flowise:2.0.9
Remember not to use variables in the nodes that are interpolated with double curly braces {{ }}, because they will not be resolved. Use the variables that are defined in the state.
I will publish the code later in my github repository if it helps you.
Regards

**Reply by vsalbuq** [reply]
Thank you, @bernejorge !

**Reply by bernejorge** [reply]
Hi man. I made some stress test to the new version, and the high ram consumption seems solved!

**Reply by cuca24** [reply]
I have memory leak with each new API request to flowise. 

@bernejorge when i use your image i get this error from the container: /bin/sh: 1: flowise: not found

**Reply by bernejorge** [reply]
Hi man. Try this one docker pull bernejorge/flowise:2.8.01
The changes I made were to make the deepClone run in a separate thread (only where possible). This way, the loopEvent doesn't get blocked. I added a cache for sequential agents. Just in case, don't use variables in prompts that aren't defined in the state.

**Reply by cuca24** [reply]
I tried but i still get error from container: /bin/sh: 1: flowise: not found

**Comment by bernejorge** [comment]
Resource consumption now

![image](https://github.com/user-attachments/assets/fb5fb562-16ae-4033-90da-77144396ba4b)
